proxy/fdleak_test.go

```diff
@@ -0,0 +1,489 @@
+		if err != nil {
+			return nil, err
+		}
+
+		// Don't capture the fd used to list /proc/self/fd
+		if fd == int(root.Fd()) {
+			continue
+		}
+
+		flags, err := fcntl(fd, syscall.F_GETFL, 0)
+		if err != nil {
+			return nil, err
+		}
+
+		fdFlags, err := fcntl(fd, syscall.F_GETFD, 0)
+		if err != nil {
+			return nil, err
+		}
+
+		// readlink on /proc/self/fd gives nice textual information
+		// about the fd
+		rl, err := os.Readlink(fmt.Sprintf("%s/%d", selfFdPath, fd))
+		if err != nil {
+			return nil, err
+		}
+
+		snap.Fds = append(snap.Fds, FdInfo{
+			Fd:          fd,
+			Flags:       flags,
+			CloseOnExec: fdFlags&syscall.FD_CLOEXEC != 0,
+			Text:        rl,
+		})
+	}
+
+	sort.Sort(ByFd(snap.Fds))
+
+	return
+}
+
+// Compare writes to w the differences between the a and b snaphots.
+// true will be returned if the two snapshots are equal, false otherwise.
+func (d *FdLeakDetector) Compare(w io.Writer, a, b *FdSnapshot) bool {
+	equal := true
+
+	i := 0
+	j := 0
+
+	for i < len(a.Fds) && j < len(b.Fds) {
+		aInfo := a.FdInfo(i)
+		bInfo := b.FdInfo(j)
+
+		if aInfo.Fd == bInfo.Fd {
+			// File descriptor found in both snapshots
+			equal = aInfo.equal(bInfo)
+			i++
+			j++
+
+			if !equal {
+				fmt.Fprintf(w, "- fd %d\n", aInfo.Fd)
+				aInfo.dump(w)
+				fmt.Fprintf(w, "+ fd %d\n", bInfo.Fd)
+				bInfo.dump(w)
+			}
+		} else if aInfo.Fd < bInfo.Fd {
+			// File descriptor present in a but not in b
+			equal = false
+			i++
+
+			fmt.Fprintf(w, "- fd %d\n", aInfo.Fd)
+			aInfo.dump(w)
+		} else {
+			// File descriptor present in b but not in a
+			equal = false
+			j++
+
+			fmt.Fprintf(w, "+ fd %d\n", bInfo.Fd)
+			bInfo.dump(w)
+		}
+	}
+
+	for ; i < len(a.Fds); i++ {
+		// File descriptor present in a but not in b
+		equal = false
+		aInfo := a.FdInfo(i)
+
+		fmt.Fprintf(w, "- fd %d\n", aInfo.Fd)
+		aInfo.dump(w)
+	}
+
+	for ; j < len(b.Fds); j++ {
+		// File descriptor present in b but not in a
+		equal = false
+		bInfo := b.FdInfo(j)
+
+		fmt.Fprintf(w, "+ fd %d\n", bInfo.Fd)
+		bInfo.dump(w)
+	}
+
+	return equal
+}
+
+func TestFdDetectorNoLeak(t *testing.T) {
+	detector := NewFdLeadDetector()
+
+	old, err := detector.Snapshot()
+	if err != nil {
+		t.Error(err)
+	}
+
+	new, err := detector.Snapshot()
+	if err != nil {
+		t.Error(err)
+	}
+
+	buffer := bytes.NewBuffer(nil)
+	equal := detector.Compare(buffer, old, new)
+	if buffer.Len() != 0 {
+		t.Fatal()
+	}
+	if !equal {
+		t.Fatal()
+	}
+}
+
+func TestFdDetectorLeak(t *testing.T) {
+	detector := NewFdLeadDetector()
+
+	old, err := detector.Snapshot()
+	if err != nil {
+		t.Error(err)
+	}
+
+	_, err = os.Open("/dev/null")
+	if err != nil {
+		t.Error(err)
+	}
+
+	new, err := detector.Snapshot()
+	if err != nil {
+		t.Error(err)
+	}
+
+	buffer := bytes.NewBuffer(nil)
+	equal := detector.Compare(buffer, old, new)
+	if equal {
+		t.Fatal()
+	}
+}
+
+func TestFdDetectorCompare(t *testing.T) {
+	tests := []struct {
+		old, new *FdSnapshot
+		equal    bool
+	}{
+		// Same fds
```

Comment:
I don't this needs to block the PR, but this static slice is huge. I'd be tempted to define individual tests such that you could then define `tests` as:

```
type snapshotTest struct {
    old, new *FdSnapshot
    equal bool
}

tests := []snapshotTest = {
    test_same_fds,
    test_same_fds_different_flags,
    test_same_fds_different_cloexe,
      :
}
```

That would make the file easier to maintain and would also mean you could optionally drop the comments as the tests would document themselves.